### PR TITLE
Update WordPress dependencies

### LINKED ISSUE #35497: FSE: see if we need package.json dependencies defined for packages that are externalized
Packages defined in https://github.com/Automattic/wp-calypso/blob/ea38075bf82481610990be449008c1fd5f8584e7/apps/full-site-editing/package.json#L32 are externalized to `window.wp.packageName` usages using the following webpack plugin
https://github.com/Automattic/wp-calypso/blob/ea38075bf82481610990be449008c1fd5f8584e7/packages/calypso-build/webpack.config.js#L18

This means that versions are not pinned (uses the host's provided dependencies instead) and that renovate PRs like the following are not needed.

https://github.com/Automattic/wp-calypso/pull/35216

Let's see if we can remove externalized package.json dependences for full site editing. Folks may want to chat with @jsnajdr, @sirreal or @ockham when folks pick this one up.

### PR DESCRIPTION
**Pull request by jsnajdr**
Big cleanup of `@wordpress` dependencies in Calypso monorepo:

In `full-site-editing` and `wpcom-block-editor`, remove `@wordpress` dependencies and other dependencies that are externalized by the `dependency-extraction-webpack-plugin` (`lodash`, `jquery`, `react`, ...) from `package.json`. These don't need to be installed in `node_modules`, because they are never accessed from there. Keep installing only the deps that are actually bundled (`classnames`, `debug`). This change also prevents Renovate Bot from submitting nonsensical "pin dependencies" PR like #35216.

This PR also removes `full-site-editing/package-lock.json`: it's not needed -- all locking is done in the root of the monorepo.

In Calypso, upgrade all `@wordpress` dependencies (we actually bundle them there) to the latest working version (we can't upgrade to the very latest ones because of WordPress/gutenberg#17165). And enumerate them all, even the transitive ones, to ensure that we don't get duplicate copies of packages bundled with different versions.

Last but not least, pin the `@wordpress/dependency-extraction-webpack-plugin`. That should make #35216 completely obsolete for now.

**How to test:**
- verify that `full-site-editing` and `wpcom-block-editor` build correctly and work correctly
- verify that Calypso is OK and that the Duplicate Package Checker webpack plugin no longer warns about `@wordpress` packages.

Fixes #35497.

### CODE REVIEW

**Comment by simison** [comment]
Yasss! Let's do this.

Only that now Eslint warns about `import/no-extraneous-dependencies` rule. :-(

**Comment by jsnajdr** [comment]
> Only that now Eslint warns about `import/no-extraneous-dependencies` rule.

In the FSE sources? Or somewhere else? Hopefully we can teach the rule about the externalized deps 🙂

**Comment by simison** [comment]
>In the FSE sources? 

Yeah:

```
npx eslint apps/wpcom-block-editor/ apps/full-site-editing/
```

image

**Comment by simison** [comment]
Does this affect folks working with TypeScript? cc @Automattic/type-review

**Comment by jsnajdr** [comment]
> Does this affect folks working with TypeScript?

That's an interesting question. We don't bundle the code, but need the types. The types are not in the `@wordpress/*` packages themselves, but rather in the `@types` scope (DefinitelyTyped repo). Does any tool break if the code is not present in `package.json` and `node_modules`?

**Comment by simison** [comment]
There might've been talks about bringing Types to `@wordpress/*` packages if I remember correctly. Someone else might fill in with links.

>Does any tool break if the code is not present in package.json and node_modules?

...and if so, would it help to add them to `optionalDependencies` with `*` range?

**Comment by noahtallen** [comment]
I pulled this branch, and it looks like full site editing is building as expected :)

The lint issue would definitely be annoying though. I tried this solution: https://github.com/benmosher/eslint-plugin-import/issues/458#issuecomment-468235671 but Eslint didn't accept it as a valid rule.

**Comment by mmtr** [comment]
Tested that `wpcom-block-editor` builds are still generated as expected 👍 

> > Only that now Eslint warns about `import/no-extraneous-dependencies` rule.
> 
> In the FSE sources? Or somewhere else? Hopefully we can teach the rule about the externalized deps 🙂

Seems there are no [options](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-extraneous-dependencies.md#options) for ignoring specific dependencies, so we'll need to disable the rule individually on each import by adding the `// eslint-disable-line import/no-extraneous-dependencies` comment.

**Comment by noahtallen** [comment]
>  so we'll need to disable the rule individually on each import

I feel like there's got to be a better way 😅 That would be a lot of lines to disable!!

File: packages/calypso-build/package.json [hunk 1]
```diff
@@ -45,7 +45,7 @@
 		"@babel/preset-typescript": "7.3.3",
 		"@wordpress/babel-plugin-import-jsx-pragma": "2.3.0",
 		"@wordpress/browserslist-config": "2.3.0",
-		"@wordpress/dependency-extraction-webpack-plugin": "^1.0.1",
+		"@wordpress/dependency-extraction-webpack-plugin": "1.1.0",
```
Comment (simison) [inline packages/calypso-build/package.json]: If you get blocked by eslint issue with removing deps, this change alone could be in separate PR.

**Comment by sirreal** [comment]
> This PR also removes `full-site-editing/package-lock.json`: it's not needed -- all locking is done in the root of the monorepo.

👍 I think this is accuray, I don't believe that package-lock file had any impact with the we install packages at this time, but it does raise an interesting point. We don't have `package-lock` in `packages/*` or `apps/*`, and it's working fine. However `packages` are part of the calypso dependency graph, whereas `apps` like `full-site-editing` live outside it. That means that `npm install` will reach dependencies of packages and install them, but not dependencies of isolated apps like `full-site-editing`. This is fine because the dependencies happen to overlap right now and can be resolved from the repository root, but that may not continue to be the case. We're relying on this overlap for `apps` to work.

This is something that [`lerna bootstrap`](https://github.com/lerna/lerna/tree/master/commands/bootstrap) addresses, it installs the dependencies of packages in the repo. We've been avoiding it so far because it adds complexity and I believe it was much slower than `npm install`, but we may need to reconsider how apps manage their dependencies.

**Comment by sirreal** [comment]
> > Only that now Eslint warns about `import/no-extraneous-dependencies` rule.
> 
> In the FSE sources? Or somewhere else? Hopefully we can teach the rule about the externalized deps 🙂

We might be able to improve this behavior with a webpack resolver for the rule: https://github.com/benmosher/eslint-plugin-import#resolvers

**Comment by sirreal** [comment]
> > Does this affect folks working with TypeScript?
> 
> That's an interesting question. We don't bundle the code, but need the types. The types are not in the `@wordpress/*` packages themselves, but rather in the `@types` scope (DefinitelyTyped repo). Does any tool break if the code is not present in `package.json` and `node_modules`?

I don't think this breaks any tooling. TypeScript complains that it can't find the types, but it's doing a lot of complaining right now as part of our gradual typing process 🙂  TypeScript might have been able to parse the JSDoc from `@wordpress` dependencies, but I don't think that's a concern.

I don't think we're using `@wordpress` dependencies in any typed files at this time anyways. I'd suggest that folks can add just the types as `devDependencies` when and if they desire:

```bash
npm i -D npm install --save @types/wordpress__{blocks,components,editor,i18n,element}
```

**Comment by sirreal** [comment]
There seem to be some regular e2e failures. I'm not sure whether this branch regresses or they were preexisting.

**Comment by jsnajdr** [comment]
> We might be able to improve this behavior with a webpack resolver for the rule:

@sirreal I tried to declare `@wordpress/data` as a core module that's part of the platform (which is true here, because it's part of the enviroment where a WordPress plugin runs):
```
settings: {
  'import/core-modules': [
    '@wordpress/data'
  ]
}
```
But then the `import/no-nodejs-modules` started complaining about importing core modules 🤷‍♂ 

I could probably fix that with the `allow` option, but then I'm in a position where I'm duplicating the list of external modules at three places rather than doing anything useful.

What kind of bugs did the `import/*` rules help us fix? Most of the time I encounter them, they don't report any real issue and just produce noise. Like in this PR.

**Comment by jsnajdr** [comment]
> There might've been talks about bringing Types to `@wordpress/*` packages if I remember correctly. Someone else might fill in with links.

Then we'll have to add them back as dev dependencies at least.

> ...and if so, would it help to add them to `package.json`'s `optionalDependencies` with `*` version range?

That would make the packages appear in the `node_modules` folder, but it's not what `optionalDependencies` are for. "optional" means that the dependency can fail to install (e.g., the package is not available for given platform) and the project can handle that and continue working nevertheless.

**Comment by jsnajdr** [comment]
> I think this is accurate, I don't believe that package-lock file had any impact with the we install packages at this time, but it does raise an interesting point. We don't have package-lock in packages/* or apps/*, and it's working fine. However packages are part of the calypso dependency graph, whereas apps like full-site-editing live outside it.

@sirreal you're totally right! We don't need lockfiles in `packages/*`, because their dependencies are installed and locked transitively when doing `npm install` in the root directory and following the `file:./packages/...` links.

But projects in `apps/*` are root projects that need to do their own `npm install`. That would become completely obvious at the moment when Calypso app gets moved to `apps/calypso`. Then there's no root `package.json` or `node_modules` any more and all `apps/*` must rely on themselves.

Solvable with Lerna Bootstrap or Yarn Workspaces. Neither is ideal for us ☹️  

I added the package locks for `full-site-editing` and `wpcom-block-editor` back in 4fc70c7.

`apps/notifications` remains to be an offender: it doesn't declare any dependencies in its `package.json` at all, and doesn't have a lockfile. Also, it's both a standalone app (shipped to WPCOM) and a module (imported by Calypso), which doesn't make things easier. Standalone app should have a lockfile, a module shouldn't.

**Comment by simison** [comment]
>apps/notifications remains to be an offender: it doesn't declare any dependencies in its package.json at all, and doesn't have a lockfile. Also, it's both a standalone app (shipped to WPCOM) and a module (imported by Calypso), which doesn't make things easier. Standalone app should have a lockfile, a module shouldn't.

You'll need an app `notifications` which imports the module `notifications`? 😁

**Comment by sirreal** [comment]
Full e2e pass:
- (desktop) https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/22035

**Comment by sirreal** [comment]
> What kind of bugs did the `import/*` rules help us fix? Most of the time I encounter them, they don't report any real issue and just produce noise. Like in this PR.

They ensure that packages do declare their `dependencies`. In the monorepo, if a package's dependency can be resolved from the root, the package works fine. However, if the published version does not declare its dependencies properly, the package may be unusable.

I don't know whether we've actually had bugs introduced that needed to be fixed, but I know that is has helped me declare dependencies correctly for packages.

File: packages/calypso-build/package.json (hunk 1 above)
Comment (sirreal) [reply packages/calypso-build/package.json]: Do we need to pin this exact version?

**Review by sirreal — APPROVED** [review]
Full e2e now seem to be passing, this looks good to me, thanks!

Some interesting questions have come up around the lint rule and the way isolated app dependencies are installed, but they're beyond the scope of this PR.

File: packages/calypso-build/package.json (hunk 1 above)
Comment (simison) [reply packages/calypso-build/package.json]: Isn't it the latest? (unless I misunderstand your question?)

Comment (sirreal) [reply packages/calypso-build/package.json]: The question was more about removing the `^` - exact version vs. version range.

Not a big deal, it's a dependency that I can't imagine ever being duplicated into a bundle that's shipped to a browser. But the version range allows for deduplication. From memory, I don't think there have been significant changes since `1.0.1`

**Comment by jsnajdr** [comment]
> Some interesting questions have come up around the lint rule and the way isolated app dependencies are installed, but they're beyond the scope of this PR.

@sirreal what do you think about disabling the `import` rules for the FSE and Block Editor apps for now? The ESLint rules are designed either for Node apps, or for mainstream React apps that bundle everything with webpack and are self-contained.

We do WordPress JS plugins, which run in a special environment where many imports are externalized. We can figure out the externals configuration (and possibly submit some PRs to the ESLint plugin) in another PR.

> are we making this too complicated for ourselves? Should we just use the wp script notation instead, off of the window object?

@gwwar Using `import '@wordpress/foo'` and externalizing that in webpack is perfectly fine and there's nothing wrong about it. It's the ESLint rule that's not flexible enough. I'd prefer to bend the ESLint rule rather than being forced to stop things that are fine.
